bats/tests/k8s/helm-install-rancher.bats

```diff
@@ -13,12 +13,95 @@ local_setup() {
+# in rancher_chart_version (and should be restored via load_var).
+determine_chart_version() {
+    if [[ -n $RD_RANCHER_IMAGE_TAG ]]; then
+        # If a version is given, check that it's compatible.
+        assert_rancher_chart_compatible "${RD_RANCHER_IMAGE_TAG#v}" || return
+        rancher_chart_version=${RD_RANCHER_IMAGE_TAG#v}
+        save_var rancher_chart_version
+        return
+    fi
+    local default_version
+    default_version=$(rancher_image_tag)
+    default_version=${default_version#v}
+    run --separate-stderr helm search repo --versions rancher-latest/rancher --output json
+    assert_success || return
+    local versions_json=$output
+    run jq_output 'map(.version).[]'
+    assert_success || return
+    run sort --version-sort <<<"$output"
+    assert_success || return
+    local versions=$output
+    local version
+    for version in $versions; do
+        if ! semver_is_valid "$version"; then
+            continue # Skip invalid / RC versions.
+        fi
+        if semver_lt "$version" "$default_version"; then
+            continue # Skip any versions older than the default version
+        fi
+        run get_chart_unsupported_version "$version"
+        assert_success || return
+        local unsupported_version=$output
+        if semver_lt "$RD_KUBERNETES_PREV_VERSION" "$unsupported_version"; then
+            # Once we find a compatible version, use it (and don't look at the
+            # rest of the chart versions).
+            trace "$(printf "Selected rancher chart version %s (wants < %s, have %s)" \
+                "$version" "$unsupported_version" "$RD_KUBERNETES_PREV_VERSION")"
+            rancher_chart_version=$version
+            save_var rancher_chart_version
+            return
+        fi
+    done
+    skip_k3s_version
+    printf "Could not find a version of rancher-latest/rancher compatible with Kubernetes %s\n" \
+        "$RD_KUBERNETES_PREV_VERSION" |
+        fail || return
+}
+
+# Check that the rancher-latest/rancher helm chart at the given version is
+# supported on the current Kubernetes version (as determined by $RD_KUBERNETES_PREV_VERSION)
+assert_rancher_chart_compatible() {
+    local chart_version=$1
+    run get_chart_unsupported_version "$chart_version"
+    assert_success || return
+    local unsupported_version=$output
+    if semver_lte "$unsupported_version" "$RD_KUBERNETES_PREV_VERSION"; then
+        skip_k3s_version
+        printf "Rancher %s wants Kubernetes < %s, have %s" \
+            "$chart_version" "$unsupported_version" "$RD_KUBERNETES_PREV_VERSION" |
+            fail || return
+    fi
+    trace "Chart $chart_version < $unsupported_version good for $RD_KUBERNETES_PREV_VERSION"
+}
+
 deploy_rancher() {
     # TODO remove `skip_unless_host_ip` once `traefik_hostname` no longer needs it
     if is_windows; then
         skip_unless_host_ip
     fi
 
+    if ! load_var rancher_chart_version; then
```

Comment:
Should this use `local`? I know that effectively this is a `@test` block, where we don't use `local`, but I feel like we should still do it when something is written as a regular function, and could potentially be called from elsewhere.

```suggestion
    local rancher_chart_version
    if ! load_var rancher_chart_version; then
```